### PR TITLE
Create react server condition alias for next/navigation api

### LINKED ISSUE #62187: [regression] "'useContext' is not exported from 'react'" error message in latest canary, works fine in v14.1.0
### Link to the code that reproduces this issue

https://github.com/stefanprobst/issue-next-browserslist

### To Reproduce

1. clone repo, `pnpm install`
2. `pnpm run build`
3. see compilation error message:

```
./node_modules/.pnpm/next@14.1.1-canary.58_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/esm/client/components/navigation.js
Attempted import error: 'useContext' is not exported from 'react' (imported as 'useContext').

Import trace for requested module:
./node_modules/.pnpm/next@14.1.1-canary.58_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/esm/client/components/navigation.js
./node_modules/.pnpm/next@14.1.1-canary.58_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/esm/api/navigation.js
./node_modules/.pnpm/next-intl@3.7.0_next@14.1.1-canary.58_react@18.2.0/node_modules/next-intl/dist/esm/server/react-server/RequestLocale.js
./node_modules/.pnpm/next-intl@3.7.0_next@14.1.1-canary.58_react@18.2.0/node_modules/next-intl/dist/esm/server.react-server.js
./app/[locale]/opengraph-image.tsx
./node_modules/.pnpm/next@14.1.1-canary.58_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/build/webpack/loaders/next-metadata-route-loader.js?page=%2F%5Blocale%5D%2Fopengraph-image%2Froute&isDynamic=1!./app/[locale]/opengraph-image.tsx?__next_metadata_route__
./node_modules/.pnpm/next@14.1.1-canary.58_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/build/webpack/loaders/next-app-loader.js?name=app%2F%5Blocale%5D%2Fopengraph-image%2Froute&page=%2F%5Blocale%5D%2Fopengraph-image%2Froute&pagePath=private-next-app-dir%2F%5Blocale%5D%2Fopengraph-image.tsx&appDir=%2Fdata%2FDevelopment%2Fplayground%2F__tmp%2Fissue-next-browserslist%2Fapp&appPaths=%2F%5Blocale%5D%2Fopengraph-image&pageExtensions=tsx&pageExtensions=ts&pageExtensions=jsx&pageExtensions=js&basePath=&assetPrefix=&nextConfigOutput=&preferredRegion=&middlewareConfig=e30%3D!./app/[locale]/opengraph-image.tsx?__next_edge_ssr_entry__
```

### Current vs. Expected behavior

- when using the latest stable *version 14.1.0* there is no error message, see [this branch](https://github.com/stefanprobst/issue-next-browserslist/tree/latest-stable) in the issue reproduction
- when *not* using the edge runtime [here]() i also don't see an error message
- also not sure if this is a warning or error message (as `next build` does complete)

### Provide environment information

```bash
Operating System:
  Platform: linux
  Arch: x64
  Version: #38~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Thu Nov  2 18:01:13 UTC 2
Binaries:
  Node: 20.11.1
  npm: 10.2.4
  Yarn: 1.22.19
  pnpm: 8.15.3
Relevant Packages:
  next: 14.1.1-canary.58 // Latest available version is detected (14.1.1-canary.58).
  eslint-config-next: N/A
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.3.3
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

Not sure

### Which stage(s) are affected? (Select all that apply)

next build (local)

### Additional context

_No response_

<sub>[NEXT-2519](https://linear.app/vercel/issue/NEXT-2519/[regression]-usecontext-is-not-exported-from-react-error-message-in)</sub>

### PR DESCRIPTION
**Pull request by huozhi**
### What

Introduce a `react-server` export condition of `next/navigation`, which only take effects in RSC layer. And it will only contain `notFound` and `redirect` related APIs, which can be shared in both server components and client components environment. This export excludes those APIs working with React context which are only working in client components. 

### Why

We fixed an issue bad alias for react-server condition of react itself in https://github.com/vercel/next.js/pull/61522/files#diff-ecb951c8d26893f6d1e4425a873b399d52346ef63eb90fba79d980cef2fabe8cL35 , this was a good fix. But we found that if you're using edge runtime with `next/navigation` it will error with bundling that you're attempted to import some client component hooks such as `useContext` from react.

So we introduced a `react-server` version of `next/navigation` that doesn't interoplate with any client hooks, can we'll bundle that one instead of original `next/navigation` when you're using it in server components or app routes.

Closes NEXT-2583
Closes NEXT-2519
Fixes #62187

### CODE REVIEW

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 08630b8e88645d5a8dd721113e0b38a32845d000

`pnpm test-dev test/e2e/app-dir/app-routes/app-custom-routes.test.ts` (PPR)
- app-custom-routes > hooks > notFound > can respond correctly
- app-custom-routes > hooks > notFound > can respond correctly (edge)

<details>
<summary>Expand output</summary>

  ● app-custom-routes › hooks › notFound › can respond correctly

    expect(received).toBeEmpty()

    Expected value to be empty received:
      "<!DOCTYPE html><html id=\"__next_error__\"><head><meta charSet=\"utf-8\"/><meta name=\"viewport\" content=\"width=device-width, initial-scale=1\"/><link rel=\"preload\" as=\"script\" fetchPriority=\"low\" href=\"/_next/static/chunks/webpack.js\"/><script src=\"/_next/static/chunks/main-app.js\" async=\"\"></script><script src=\"/_next/static/chunks/app-pages-internals.js\" async=\"\"></script><meta name=\"robots\" content=\"noindex\"/><meta name=\"next-error\" content=\"not-found\"/><title>Next.js</title><meta name=\"description\" content=\"Generated by Next.js\"/><script src=\"/_next/static/chunks/polyfills.js\" noModule=\"\"></script></head><body><script src=\"/_next/static/chunks/webpack.js\" async=\"\"></script><script>(self.__next_f=self.__next_f||[]).push([0]);self.__next_f.push([2,null])</script><script>self.__next_f.push([1,\"0:D{\\\"name\\\":\\\"rj\\\",\\\"env\\\":\\\"server\\\"}\\n\"])</script><script>self.__next_f.push([1,\"1:I[\\\"(app-pages-browser)/./node_modules/.pnpm/file+..+next-repo-d5e6308d80e3bee0943ebc5d7ca1d3aca63cbe8c3b980d0bafcc7169fea5c267+packages+n_newlaz23hq4lkb6zszgiinl2ti/node_modules/next/dist/client/components/app-router.js\\\",[\\\"app-pages-internals\\\",\\\"static/chunks/app-pages-internals.js\\\"],\\\"\\\"]\\n5:I[\\\"(app-pages-browser)/./node_modules/.pnpm/file+..+next-repo-d5e6308d80e3bee0943ebc5d7ca1d3aca63cbe8c3b980d0bafcc7169fea5c267+packages+n_newlaz23hq4lkb6zszgiinl2ti/node_modules/next/dist/client/components/layout-router.\"])</script><script>self.__next_f.push([1,\"js\\\",[\\\"app-pages-internals\\\",\\\"static/chunks/app-pages-internals.js\\\"],\\\"\\\"]\\n6:I[\\\"(app-pages-browser)/./node_modules/.pnpm/file+..+next-repo-d5e6308d80e3bee0943ebc5d7ca1d3aca63cbe8c3b980d0bafcc7169fea5c267+packages+n_newlaz23hq4lkb6zszgiinl2ti/node_modules/next/dist/client/components/render-from-template-context.js\\\",[\\\"app-pages-internals\\\",\\\"static/chunks/app-pages-internals.js\\\"],\\\"\\\"]\\n9:I[\\\"(app-pages-browser)/./node_modules/.pnpm/file+..+next-repo-d5e6308d80e3bee0943ebc5d7ca1d3aca63cbe8c3b980d0bafcc7169fea5c267+pack\"])</script><script>self.__next_f.push([1,\"ages+n_newlaz23hq4lkb6zszgiinl2ti/node_modules/next/dist/client/components/error-boundary.js\\\",[\\\"app-pages-internals\\\",\\\"static/chunks/app-pages-internals.js\\\"],\\\"\\\"]\\n2:D{\\\"name\\\":\\\"\\\",\\\"env\\\":\\\"server\\\"}\\n3:D{\\\"name\\\":\\\"ParallelRouteDefault\\\",\\\"env\\\":\\\"server\\\"}\\n4:D{\\\"name\\\":\\\"RootLayout\\\",\\\"env\\\":\\\"server\\\"}\\n7:D{\\\"name\\\":\\\"NotFound\\\",\\\"env\\\":\\\"server\\\"}\\n\"])</script><script>self.__next_f.push([1,\"7:[[\\\"$\\\",\\\"title\\\",null,{\\\"children\\\":\\\"404: This page could not be found.\\\"}],[\\\"$\\\",\\\"div\\\",null,{\\\"style\\\":{\\\"fontFamily\\\":\\\"system-ui,\\\\\\\"Segoe UI\\\\\\\",Roboto,Helvetica,Arial,sans-serif,\\\\\\\"Apple Color Emoji\\\\\\\",\\\\\\\"Segoe UI Emoji\\\\\\\"\\\",\\\"height\\\":\\\"100vh\\\",\\\"textAlign\\\":\\\"center\\\",\\\"display\\\":\\\"flex\\\",\\\"flexDirection\\\":\\\"column\\\",\\\"alignItems\\\":\\\"center\\\",\\\"justifyContent\\\":\\\"center\\\"},\\\"children\\\":[\\\"$\\\",\\\"div\\\",null,{\\\"children\\\":[[\\\"$\\\",\\\"style\\\",null,{\\\"dangerouslySetInnerHTML\\\":{\\\"__html\\\":\\\"body{color:#000;background:#fff;margin:0}.next-error-h1{border-right:1px solid rgba(0,0,0,.3)}@media (prefers-color-scheme:dark){body{color:#fff;background:#000}.next-error-h1{border-right:1px solid rgba(255,255,255,.3)}}\\\"}}],[\\\"$\\\",\\\"h1\\\",null,{\\\"className\\\":\\\"next-error-h1\\\",\\\"style\\\":{\\\"display\\\":\\\"inline-block\\\",\\\"margin\\\":\\\"0 20px 0 0\\\",\\\"padding\\\":\\\"0 23px 0 0\\\",\\\"fontSize\\\":24,\\\"fontWeight\\\":500,\\\"verticalAlign\\\":\\\"top\\\",\\\"lineHeight\\\":\\\"49px\\\"},\\\"children\\\":\\\"404\\\"}],[\\\"$\\\",\\\"div\\\",null,{\\\"style\\\":{\\\"display\\\":\\\"inline-block\\\"},\\\"children\\\":[\\\"$\\\",\\\"h2\\\",null,{\\\"style\\\":{\\\"fontSize\\\":14,\\\"fontWeight\\\":400,\\\"lineHeight\\\":\\\"49px\\\",\\\"margin\\\":0},\\\"children\\\":\\\"This page could not be found.\\\"}]}]]}]}]]\\n\"])</script><script>self.__next_f.push([1,\"4:[\\\"$\\\",\\\"html\\\",null,{\\\"lang\\\":\\\"en\\\",\\\"children\\\":[\\\"$\\\",\\\"body\\\",null,{\\\"children\\\":[\\\"$\\\",\\\"$L5\\\",null,{\\\"parallelRouterKey\\\":\\\"children\\\",\\\"segmentPath\\\":[\\\"children\\\"],\\\"loading\\\":\\\"$undefined\\\",\\\"loadingStyles\\\":\\\"$undefined\\\",\\\"loadingScripts\\\":\\\"$undefined\\\",\\\"hasLoading\\\":false,\\\"error\\\":\\\"$undefined\\\",\\\"errorStyles\\\":\\\"$undefined\\\",\\\"errorScripts\\\":\\\"$undefined\\\",\\\"template\\\":[\\\"$\\\",\\\"$L6\\\",null,{}],\\\"templateStyles\\\":\\\"$undefined\\\",\\\"templateScripts\\\":\\\"$undefined\\\",\\\"notFound\\\":\\\"$7\\\",\\\"notFoundStyles\\\":[],\\\"styles\\\":null}]}]}]\\n8:D{\\\"name\\\":\\\"\\\",\\\"env\\\":\\\"server\\\"}\\na:[\\\"childr\"])</script><script>self.__next_f.push([1,\"en\\\"]\\n0:[null,[\\\"$\\\",\\\"$L1\\\",null,{\\\"buildId\\\":\\\"development\\\",\\\"assetPrefix\\\":\\\"\\\",\\\"initialCanonicalUrl\\\":\\\"/hooks/not-found\\\",\\\"initialTree\\\":[\\\"__NOT_FOUND__\\\",{\\\"children\\\":[\\\"__DEFAULT__\\\",{}]},\\\"$undefined\\\",\\\"$undefined\\\",true],\\\"initialSeedData\\\":[\\\"\\\",{\\\"children\\\":[\\\"__DEFAULT__\\\",{},[[\\\"$L2\\\",\\\"$L3\\\"],null]]},[\\\"$4\\\",null]],\\\"couldBeIntercepted\\\":false,\\\"initialHead\\\":[[\\\"$\\\",\\\"meta\\\",null,{\\\"name\\\":\\\"robots\\\",\\\"content\\\":\\\"noindex\\\"}],\\\"$L8\\\"],\\\"globalErrorComponent\\\":\\\"$9\\\",\\\"missingSlots\\\":\\\"$Wa\\\"}]]\\n\"])</script><script>self.__next_f.push([1,\"3:E{\\\"digest\\\":\\\"NEXT_NOT_FOUND\\\",\\\"message\\\":\\\"NEXT_NOT_FOUND\\\",\\\"stack\\\":\\\"Error: NEXT_NOT_FOUND\\\\n    at notFound (webpack-internal:///(rsc)/./node_modules/.pnpm/file+..+next-repo-d5e6308d80e3bee0943ebc5d7ca1d3aca63cbe8c3b980d0bafcc7169fea5c267+packages+n_newlaz23hq4lkb6zszgiinl2ti/node_modules/next/dist/client/components/not-found.js:23:19)\\\\n    at ParallelRouteDefault (webpack-internal:///(rsc)/./node_modules/.pnpm/file+..+next-repo-d5e6308d80e3bee0943ebc5d7ca1d3aca63cbe8c3b980d0bafcc7169fea5c267+packages+n_newlaz23hq4lkb6zszgiinl2ti/node_modules/next/dist/client/components/parallel-route-default.js:23:28)\\\\n    at eA (/tmp/next-install-16ce669a707af76586c10ed82fff7fc7f018730736399421462259b1d57af9c8/node_modules/.pnpm/file+..+next-repo-d5e6308d80e3bee0943ebc5d7ca1d3aca63cbe8c3b980d0bafcc7169fea5c267+packages+n_newlaz23hq4lkb6zszgiinl2ti/node_modules/next/dist/compiled/next-server/app-page-experimental.runtime.dev.js:35:367656)\\\\n    at e (/tmp/next-install-16ce669a707af76586c10ed82fff7fc7f018730736399421462259b1d57af9c8/node_modules/.pnpm/file+..+next-repo-d5e6308d80e3bee0943ebc5d7ca1d3aca63cbe8c3b980d0bafcc7169fea5c267+packages+n_newlaz23hq4lkb6zszgiinl2ti/node_modules/next/dist/compiled/next-server/app-page-experimental.runtime.dev.js:35:372580)\\\\n    at eG (/tmp/next-install-16ce669a707af76586c10ed82fff7fc7f018730736399421462259b1d57af9c8/node_modules/.pnpm/file+..+next-repo-d5e6308d80e3bee0943ebc5d7ca1d3aca63cbe8c3b980d0bafcc7169fea5c267+packages+n_newlaz23hq4lkb6zszgiinl2ti/node_modules/next/dist/compiled/next-server/app-page-experimental.runtime.dev.js:35:373077)\\\\n    at e8 (/tmp/next-install-16ce669a707af76586c10ed82fff7fc7f018730736399421462259b1d57af9c8/node_modules/.pnpm/file+..+next-repo-d5e6308d80e3bee0943ebc5d7ca1d3aca63cbe8c3b980d0bafcc7169fea5c267+packages+n_newlaz23hq4lkb6zszgiinl2ti/node_modules/next/dist/compiled/next-server/app-page-experimental.runtime.dev.js:35:379436)\\\\n    at eM (/tmp/next-install-16ce669a707af76586c10ed82fff7fc7f018730736399421462259b1d57af9c8/node_modules/.pnpm/file+..+next-repo-d5e6308d80e3bee0943ebc5d7ca1d3aca63cbe8c3b980d0bafcc7169fea5c267+packages+n_newlaz23hq4lkb6zszgiinl2ti/node_modules/next/dist/compiled/next-server/app-page-experimental.runtime.dev.js:35:368694)\\\\n    at eA (/tmp/next-install-16ce669a707af76586c10ed82fff7fc7f018730736399421462259b1d57af9c8/node_modules/.pnpm/file+..+next-repo-d5e6308d80e3bee0943ebc5d7ca1d3aca63cbe8c3b980d0bafcc7169fea5c267+packages+n_newlaz23hq4lkb6zszgiinl2ti/node_modules/next/dist/compiled/next-server/app-page-experimental.runtime.dev.js:35:367526)\\\\n    at e (/tmp/next-install-16ce669a707af76586c10ed82fff7fc7f018730736399421462259b1d57af9c8/node_modules/.pnpm/file+..+next-repo-d5e6308d80e3bee0943ebc5d7ca1d3aca63cbe8c3b980d0bafcc7169fea5c267+packages+n_newlaz23hq4lkb6zszgiinl2ti/node_modules/next/dist/compiled/next-server/app-page-experimental.runtime.dev.js:35:372580)\\\\n    at eG (/tmp/next-install-16ce669a707af76586c10ed82fff7fc7f018730736399421462259b1d57af9c8/node_modules/.pnpm/file+..+next-repo-d5e6308d80e3bee0943ebc5d7ca1d3aca63cbe8c3b980d0bafcc7169fea5c267+packages+n_newlaz23hq4lkb6zszgiinl2ti/node_modules/next/dist/compiled/next-server/app-page-experimental.runtime.dev.js:35:373077)\\\\n    at /tmp/next-install-16ce669a707af76586c10ed82fff7fc7f018730736399421462259b1d57af9c8/node_modules/.pnpm/file+..+next-repo-d5e6308d80e3bee0943ebc5d7ca1d3aca63cbe8c3b980d0bafcc7169fea5c267+packages+n_newlaz23hq4lkb6zszgiinl2ti/node_modules/next/dist/compiled/next-server/app-page-experimental.runtime.dev.js:35:369732\\\\n    at Array.toJSON (/tmp/next-install-16ce669a707af76586c10ed82fff7fc7f018730736399421462259b1d57af9c8/node_modules/.pnpm/file+..+next-repo-d5e6308d80e3bee0943ebc5d7ca1d3aca63cbe8c3b980d0bafcc7169fea5c267+packages+n_newlaz23hq4lkb6zszgiinl2ti/node_modules/next/dist/compiled/next-server/app-page-experimental.runtime.dev.js:35:370345)\\\\n    at stringify (\\u003canonymous\\u003e)\\\\n    at e8 (/tmp/next-install-16ce669a707af76586c10ed82fff7fc7f018730736399421462259b1d57af9c8/node_modules/.pnpm/file+..+next-repo-d5e6308d80e3bee0943ebc5d7ca1d3aca63cbe8c3b980d0bafcc7169fea5c267+packages+n_newlaz23hq4lkb6zszgiinl2ti/node_modules/next/dist/compiled/next-server/app-page-experimental.runtime.dev.js:35:379535)\\\\n    at e5 (/tmp/next-install-16ce669a707af76586c10ed82fff7fc7f018730736399421462259b1d57af9c8/node_modules/.pnpm/file+..+next-repo-d5e6308d80e3bee0943ebc5d7ca1d3aca63cbe8c3b980d0bafcc7169fea5c267+packages+n_newlaz23hq4lkb6zszgiinl2ti/node_modules/next/dist/compiled/next-server/app-page-experimental.runtime.dev.js:35:380150)\\\\n    at Timeout._onTimeout (/tmp/next-install-16ce669a707af76586c10ed82fff7fc7f018730736399421462259b1d57af9c8/node_modules/.pnpm/file+..+next-repo-d5e6308d80e3bee0943ebc5d7ca1d3aca63cbe8c3b980d0bafcc7169fea5c267+packages+n_newlaz23hq4lkb6zszgiinl2ti/node_modules/next/dist/compiled/next-server/app-page-experimental.runtime.dev.js:35:368854)\\\\n    at listOnTimeout (node:internal/timers:573:17)\\\\n    at process.processTimers (node:internal/timers:514:7)\\\"}\\n\"])</script><script>self.__next_f.push([1,\"b:\\\"$Sreact.fragment\\\"\\nc:[[\\\"$\\\",\\\"meta\\\",\\\"0\\\",{\\\"name\\\":\\\"viewport\\\",\\\"content\\\":\\\"width=device-width, initial-scale=1\\\"}],[\\\"$\\\",\\\"meta\\\",\\\"1\\\",{\\\"charSet\\\":\\\"utf-8\\\"}],[\\\"$\\\",\\\"title\\\",\\\"2\\\",{\\\"children\\\":\\\"Next.js\\\"}],[\\\"$\\\",\\\"meta\\\",\\\"3\\\",{\\\"name\\\":\\\"description\\\",\\\"content\\\":\\\"Generated by Next.js\\\"}]]\\n8:[\\\"$\\\",\\\"$b\\\",\\\"NR75TJTl_MyC5Dn82o0Aa\\\",{\\\"children\\\":\\\"$c\\\"}]\\n2:null\\n\"])</script></body></html>"

      480 |
      481 |           expect(res.status).toEqual(404)
    > 482 |           expect(await res.text()).toBeEmpty()
          |                                    ^
      483 |         })
      484 |
      485 |         it('can respond correctly (edge)', async () => {

      at Object.toBeEmpty (e2e/app-dir/app-routes/app-custom-routes.test.ts:482:36)

  ● app-custom-routes › hooks › notFound › can respond correctly (edge)

    Tests cannot be nested. Test "can respond correctly" cannot run because it is nested within "can respond correctly (edge)".

      484 |
      485 |         it('can respond correctly (edge)', async () => {
    > 486 |           it('can respond correctly', async () => {
          |           ^
      487 |             const res = await next.fetch(basePath + '/hooks/not-found/edge')
      488 |
      489 |             expect(res.status).toEqual(404)

      at eventHandler (../node_modules/.pnpm/jest-circus@29.7.0/node_modules/jest-circus/build/eventHandler.js:134:11)
      at Object.it (e2e/app-dir/app-routes/app-custom-routes.test.ts:486:11)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js add-rsc-version-navigation | Change |
| - | - | - | - |
| buildDuration | 13.8s | 13.9s | ⚠️ +157ms |
| buildDurationCached | 7.5s | 6.4s | N/A |
| nodeModulesSize | 197 MB | 197 MB | ⚠️ +32.9 kB |
| nextStartRea..uration (ms) | 408ms | 405ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js add-rsc-version-navigation | Change |
| - | - | - | - |
| 1068-HASH.js gzip | 30.3 kB | 30.4 kB | N/A |
| 3f784ff6-HASH.js gzip | 53.7 kB | 53.7 kB | N/A |
| 4944-HASH.js gzip | 5.04 kB | 5.04 kB | N/A |
| 8423.HASH.js gzip | 181 B | 181 B | ✓ |
| framework-HASH.js gzip | 45.2 kB | 45.2 kB | ✓ |
| main-app-HASH.js gzip | 241 B | 241 B | ✓ |
| main-HASH.js gzip | 32.2 kB | 32.2 kB | N/A |
| webpack-HASH.js gzip | 1.7 kB | 1.7 kB | N/A |
| Overall change | 45.6 kB | 45.6 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js add-rsc-version-navigation | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js add-rsc-version-navigation | Change |
| - | - | - | - |
| _app-HASH.js gzip | 196 B | 197 B | N/A |
| _error-HASH.js gzip | 184 B | 184 B | ✓ |
| amp-HASH.js gzip | 503 B | 505 B | N/A |
| css-HASH.js gzip | 323 B | 325 B | N/A |
| dynamic-HASH.js gzip | 2.5 kB | 2.5 kB | ✓ |
| edge-ssr-HASH.js gzip | 258 B | 258 B | ✓ |
| head-HASH.js gzip | 353 B | 352 B | N/A |
| hooks-HASH.js gzip | 370 B | 371 B | N/A |
| image-HASH.js gzip | 4.21 kB | 4.2 kB | N/A |
| index-HASH.js gzip | 259 B | 259 B | ✓ |
| link-HASH.js gzip | 2.68 kB | 2.67 kB | N/A |
| routerDirect..HASH.js gzip | 313 B | 312 B | N/A |
| script-HASH.js gzip | 386 B | 386 B | ✓ |
| withRouter-HASH.js gzip | 309 B | 309 B | ✓ |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 4.01 kB | 4.01 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js add-rsc-version-navigation | Change |
| - | - | - | - |
| _buildManifest.js gzip | 485 B | 485 B | ✓ |
| Overall change | 485 B | 485 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js add-rsc-version-navigation | Change |
| - | - | - | - |
| index.html gzip | 528 B | 527 B | N/A |
| link.html gzip | 540 B | 539 B | N/A |
| withRouter.html gzip | 523 B | 521 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js add-rsc-version-navigation | Change |
| - | - | - | - |
| edge-ssr.js gzip | 94.9 kB | 94.9 kB | N/A |
| page.js gzip | 3.06 kB | 3.06 kB | ✓ |
| Overall change | 3.06 kB | 3.06 kB | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js add-rsc-version-navigation | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 625 B | 625 B | ✓ |
| middleware-r..fest.js gzip | 151 B | 151 B | ✓ |
| middleware.js gzip | 25.4 kB | 25.5 kB | N/A |
| edge-runtime..pack.js gzip | 839 B | 839 B | ✓ |
| Overall change | 1.61 kB | 1.61 kB | ✓ |

</details>

<details>
<summary><strong>Next Runtimes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js add-rsc-version-navigation | Change |
| - | - | - | - |
| app-page-exp...dev.js gzip | 170 kB | 170 kB | N/A |
| app-page-exp..prod.js gzip | 96.5 kB | 96.6 kB | N/A |
| app-page-tur..prod.js gzip | 98.3 kB | 98.3 kB | N/A |
| app-page-tur..prod.js gzip | 92.8 kB | 92.8 kB | N/A |
| app-page.run...dev.js gzip | 151 kB | 151 kB | N/A |
| app-page.run..prod.js gzip | 91.3 kB | 91.3 kB | N/A |
| app-route-ex...dev.js gzip | 22 kB | 22 kB | N/A |
| app-route-ex..prod.js gzip | 15 kB | 15 kB | N/A |
| app-route-tu..prod.js gzip | 15 kB | 15 kB | N/A |
| app-route-tu..prod.js gzip | 14.7 kB | 14.8 kB | N/A |
| app-route.ru...dev.js gzip | 21.7 kB | 21.7 kB | N/A |
| app-route.ru..prod.js gzip | 14.7 kB | 14.8 kB | N/A |
| pages-api-tu..prod.js gzip | 9.5 kB | 9.51 kB | N/A |
| pages-api.ru...dev.js gzip | 9.78 kB | 9.79 kB | N/A |
| pages-api.ru..prod.js gzip | 9.5 kB | 9.51 kB | N/A |
| pages-turbo...prod.js gzip | 22.3 kB | 22.3 kB | N/A |
| pages.runtim...dev.js gzip | 23 kB | 23 kB | N/A |
| pages.runtim..prod.js gzip | 22.3 kB | 22.3 kB | N/A |
| server.runti..prod.js gzip | 50.5 kB | 50.5 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>build cache</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js add-rsc-version-navigation | Change |
| - | - | - | - |
| 0.pack gzip | 1.56 MB | 1.56 MB | ⚠️ +5.2 kB |
| index.pack gzip | 105 kB | 105 kB | ⚠️ +427 B |
| Overall change | 1.66 MB | 1.67 MB | ⚠️ +5.62 kB |

</details>

<details>
<summary><strong>Diff details</strong></summary>

<details>
<summary>Diff for <strong>middleware.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>edge-ssr.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>1068-HASH.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page-exp..ntime.dev.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page-exp..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page-tur..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page-tur..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page.runtime.dev.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page.runtime.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-route-ex..ntime.dev.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-route-ex..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-route-tu..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-route-tu..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-route.runtime.dev.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-route.ru..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>pages-api-tu..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>pages-api.runtime.dev.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>pages-api.ru..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>pages-turbo...time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>pages.runtime.dev.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>pages.runtime.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>server.runtime.prod.js</strong></summary>

Diff too large to display
</details>

</details>

</details>
Commit: 08630b8e88645d5a8dd721113e0b38a32845d000

File: packages/next/src/build/create-compiler-aliases.ts [hunk 1]
```diff
@@ -220,12 +220,16 @@ export function createNextApiEsmAliases() {
...
-export function createAppRouterApiAliases() {
-  const mapping = {
+export function createAppRouterApiAliases(isServerOnlyLayer: boolean) {
+  const mapping: Record<string, string> = {
     head: 'next/dist/client/components/noop-head',
     dynamic: 'next/dist/api/app-dynamic',
   }
 
+  if (isServerOnlyLayer) {
+    mapping['navigation'] = 'next/dist/api/navigation.react-server'
```
Comment (kwonoj) [inline packages/next/src/build/create-compiler-aliases.ts:230]: should we have same for the turbopack? or it just passes new test automatically?

Comment (huozhi) [reply packages/next/src/build/create-compiler-aliases.ts:230]: @kwonoj good call, we can add for turbopack too

**Review by gnoff — COMMENTED** [review]
One of the nice things we can do with a fork like this is just not export the invalid methods from the react server module. Instead of waiting until runtime to detect an invalid usage we can know at compile time. This also means the error won’t creep up only in some dynamic render or in a revalidate rather than during a build.

it’s unfortunate that the error message itself would be more generic (really it’s exactly the error that instigated this fix in the first place for useContext exporting from react). However I thought we already have some error specialization where we take a generic import error and replace it with something that provides better explanation.


Have you looked at that approach already?

**Comment by huozhi** [comment]
@gnoff I was actually using that before the last few commits, but the error message is not ideal. We also have that in SWR. You might see sth like `(0, __next_navigation_mod).usePathname is not a function`, which is pretty hard to determine the format and source import module. So I still keep the wrong time error there. This is a quite general issue that we cannot mock the exports for specific export condition.

The better approach is we do a follow up change that replaces the runtime checking with SWC imports checking in build time. This only resolves the bundling error to unblock the build, later we introduce the better build time checking so it can error earlier

**Comment by gnoff** [comment]
Makes sense!

**Review by gnoff — APPROVED** [review]

**Review by shuding — APPROVED** [review]

File: packages/next/src/build/create-compiler-aliases.ts (hunk 1 above)
Comment (huozhi) [reply packages/next/src/build/create-compiler-aliases.ts:230]: Tested it that turbopack doesn't have this issue, I'll create a separate PR to follow up the alias for turbopack

**Pull request merged**